### PR TITLE
candump down/up device monitoring

### PR DESCRIPTION
**Pull request by ajneu**
See #26

### CODE REVIEW

**Comment by ajneu** [comment]
To test: 
```bash
candump -R -D vcan0
```

In other shell change can-device down and up...:
```bash
sudo ip link set vcan0 down
sudo ip link set vcan0 up
sudo ip link set vcan0 down
sudo ip link set vcan0 up
# etc...
```

File: candump.c [hunk 1]
```diff
@@ -152,6 +155,20 @@ void sigterm(int signo)
 	running = 0;
 }
 
+// idx of "detected" devices is cached
+int is_idx_cached(int ifidx) {
+	int cached = 0;
+	int i;
+
+	for (i=0; i < MAXIFNAMES; i++) {
```
Comment (marckleinebudde) [inline candump.c]: space around "="

File: candump.c [hunk 2]
```diff
@@ -236,6 +281,16 @@ int main(int argc, char **argv)
 	struct timeval tv, last_tv;
 	struct timeval timeout, timeout_config = { 0, 0 }, *timeout_current = NULL;
 	FILE *logfile = NULL;
+	struct netlink_struc {
+		int s;
+		struct sockaddr_nl saddr;
+	} netlink_s = {
+			.s = socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE),
```
Comment (marckleinebudde) [inline candump.c]: Sorry for not beeing clear in the first review, better use only C99 initializers for constant types. So move the: `s = socket();` next to the `bind()`.

File: candump.c [hunk 3]
```diff
@@ -399,9 +458,17 @@ int main(int argc, char **argv)
 			silent = SILENT_OFF; /* default output */
 	}
 
+	// netlink to determine down/up
+	bind(netlink_s.s, (struct sockaddr *) &netlink_s.saddr, sizeof(netlink_s.saddr));
```
Comment (marckleinebudde) [inline candump.c]: please add error handling

Comment (marckleinebudde) [inline candump.c]: remove the space between the cast and the &netlink_s

File: candump.c [hunk 4]
```diff
@@ -635,6 +704,41 @@ int main(int argc, char **argv)
 			continue;
 		}
 
+		// check netlink socket
+		if (FD_ISSET(netlink_s.s, &rdfs)) {
+			char buf[4096];
```
Comment (marckleinebudde) [inline candump.c]: please move this into a separate function

File: candump.c [hunk 5]
```diff
@@ -635,6 +704,41 @@ int main(int argc, char **argv)
...
+				if (nh->nlmsg_type == NLMSG_ERROR)
+					continue;
+				if (nh->nlmsg_type == RTM_NEWLINK) {
+					struct ifinfomsg *rtif = NLMSG_DATA(nh);
+
+					if (is_idx_cached(rtif->ifi_index)) {
+						char ifname[IF_NAMESIZE];
+						if_indextoname(rtif->ifi_index, ifname);
+
+						const int will_exit = (down_causes_exit && !(rtif->ifi_flags&IFF_RUNNING));
```
Comment (marckleinebudde) [inline candump.c]: please move all variable declarations to the beginning of the scope. (I know, kernel C-style is not consequent in using C99 or not :)

File: candump.c [hunk 6]
```diff
@@ -647,16 +751,23 @@ int main(int argc, char **argv)
...
 
+				idx = idx2dindex(addr.can_ifindex, s[i]);
+
 				nbytes = recvmsg(s[i], &msg, 0);
 				if (nbytes < 0) {
-					if ((errno == ENETDOWN) && !down_causes_exit) {
-						fprintf(stderr, "read: %s \t\t(not exiting because of -D flag)\n", strerror(errno));
+					if (errno == ENETDOWN) {
+					    if (down_causes_exit) {
+						//if (!report_down_up) { /* no previous reporting message, so report here before exit */
```
Comment (marckleinebudde) [inline candump.c]: If you don't need the commented out stuff, remove it

**Comment by hartkopp** [comment]
I generally don't like this patch.
IMO it can make sense to omit the exit of candump as the previous implemented.
(And when doing so the output should not look like "read: Network is down 		(not exiting because of -D flag)" but "vcan0: interface gone down (temporarily?)")
But IMO this patch is just bloating the code. In which use-case does it really help?
Aka: Why should people invoke an ifdown/ifup at runtime?
And what happens when you completely remove the interface?
Do you probably want to invoke "ip link set can0 restart" instead??

File: candump.c [hunk 7]
```diff
@@ -647,16 +751,23 @@ int main(int argc, char **argv)
 				msg.msg_controllen = sizeof(ctrlmsg);  
 				msg.msg_flags = 0;
 
+				idx = idx2dindex(addr.can_ifindex, s[i]);
```
Comment (hartkopp) [inline candump.c]: Additionally to the fact that I don't like the patch addr.can_ifindex is only valid AFTER recvmsg() ... :-)

**Comment by ajneu** [comment]
> But IMO this patch is just bloating the code. In which use-case does it really help?

There's a common use-case for binaries using [libsocketcan](http://git.pengutronix.de/?p=tools/libsocketcan.git;a=summary), to pull the device down (`can_do_stop`) and back up again (`can_do_start`). It's the only way I know of, to clear the socketcan transmit buffer. (And I really want to clear pending messages from it, because if a CAN-bus reconnect only occurs in say 5 minutes, then I don't want to transmit 5-minute old frames on reconnection.) See also #26 

> Aka: Why should people invoke an ifdown/ifup at runtime?

Answered above.

> And what happens when you completely remove the interface?

Does it have a usecase?

Here's what happens:
```bash
   $./candump -D -R vcan0
# vcan0 is up

                      ## In other shell do:     sudo ip link delete vcan0
# vcan0 is down
read: No such device
```

> Do you probably want to invoke "ip link set can0 restart" instead??

Does that command exist???
```bash
   $ ip -V
ip utility, iproute2-ss160518   ## it's this one:  https://packages.debian.org/stretch/iproute2

   $ ip link help 2>&1 | grep restart
   ## no result

   $ sudo ip link set vcan0 restart
Error: either "dev" is duplicate, or "restart" is a garbage.
```

PS: But I've now rechecked libsocketcan, and found that I missed `can_do_restart` till now (I previously always did `can_do_stop(can0); can_do_start(can0)`. It seems that libsocketcan's `can_do_restart(vcan0)` does not cause down-and-up... so either it is buggy, or it does some kind of transparent restart (which would be great.). Does `can_do_restart(can0)` clear the socketcan transmit buffer?

File: candump.c (hunk 2 above)
Comment (ajneu) [inline candump.c]: I've changed it, but I'm not sure if it was, what you were aiming at...

**Comment by ajneu** [comment]
> But IMO this patch is just bloating the code.

Well I've now move the netlink parts into a separate function as suggested by Marc Kleine-Budde [here](https://github.com/linux-can/can-utils/pull/29#discussion-diff-77008181).
'hope that makes it somewhat better.

**Comment by ajneu** [comment]
@ Marc Kleine-Budde: I'm interested in suggesting some changes to upstream libsocketcan ([ref](https://github.com/ajneu/libsocketcan/commits/master)). What is the correct way of proposing changes for that project?
Thanks

**Comment by hartkopp** [comment]
> >  But IMO this patch is just bloating the code. In which use-case does it really help?

> There's a common use-case for binaries using libsocketcan, to pull the device down (can_do_stop) and back up again (can_do_start). It's the only way I know of, to clear the socketcan transmit buffer. (And I really want to clear pending messages from it, because if a CAN-bus reconnect only occurs in say 5 minutes, then I don't want to transmit 5-minute old frames on reconnection.) See also #26 

Ok. This is a use-case for doing a down/up sequence.

But do we really need to have this blabla of 'is down' and 'is up' in candump for this?
Having the information of a 'down' interface seems to be enough - as it is a new reaction instead of terminating candump. But when the interface comes up again some traffic will show up anyway.

Two additional infos:

1. 'ip link set can0 type can restart' is a way to reset the CAN controller in the case of BUS_OFF
When you define 'ip link set can0 type can restart-ms 100' the controller is restarted 100ms after BUS_OFF automatically. So it does not really help in the case of an open bus without a second node.
And I saw that 'ip link set can0 type can restart' says 'RTNETLINK answers: Device or resource busy' - hm. I thought this should work ...

2. I found this interesting document and comment https://www.coverfire.com/articles/queueing-in-the-linux-network-stack/#comment-480695 on the Internet. But even if we flush the tx queue we still would have the single CAN frame in the CAN controller.

We could think about a new  'ip link set can0 type can flush' command to flush the tx path without doing a down/up sequence.

**Comment by ajneu** [comment]
> But do we really need to have this blabla of 'is down' and 'is up' in candump for this?

Yes, I want it, because then I can see if a device is pulled down and up again. It's only active with `-R flag`.

> We could think about a new 'ip link set can0 type can flush' command to flush the tx path without doing a down/up sequence.
Yes. 
(But I'll be using down/up, since it's working in my current kernel, and I don't have to mess around with patching the kernel, etc. 
For me, a patched candump is simpler... and hope that my patch here can be accepted upstream.)

**Comment by hartkopp** [comment]
> > But do we really need to have this blabla of 'is down' and 'is up' in candump for this?

> Yes, I want it, because then I can see if a device is pulled down and up again. It's only active with -R flag.

Ok - I must admit that my question was a rhetorical one.

The idea behind Unix tools is its philosophy https://en.wikipedia.org/wiki/Unix_philosophy :
Make each program do one thing well. To do a new job, build afresh rather than complicate old programs by adding new "features".

candump is a tool to display or dump CAN data.
candump is no tool to monitor and display CAN interface changes.

I'm fine with your change to preserve candump's termination so that it behaves with explicitly named CAN interfaces in the same way as it does with the 'any' device.
But adding all these netlink stuff just to print changing interface states on stderr? No.

Following the Unix philosophy you may create a new tool that displays CAN interface states - e.g. in another terminal of your system. Or you can send CAN interface states on a virtual CAN interface and dump this interface with candump too.

Or you may have your desired functionality in your private repo.
But don't expect this netlink stuff to become part of candump.

**Comment by ajneu** [comment]
> Following the Unix philosophy you may create a new tool that displays CAN interface states - e.g. in another terminal of your system. 

I want an output that preserves temporal order of occurence to 100%.
How can I combine candump with CAN interface states?

Perhaps you though with the following, but I see it as rather obscure:

> Or you can send CAN interface states on a virtual CAN interface and dump this interface with candump too.

I thought the Unix tools and philosophy work through file-writing and file-reading. Dumping via a virtual CAN interface seems a bit far-fetched, to be honest.
Do you mean that I should invent some CAN-frame ID's and payload data to symbolize up/down etc?
That's about as far away from elegant, as I can conceive. And I'm not sure I'll have preserved temporal order.

> Or you may have your desired functionality in your private repo.

Yip. This is the only viable solution to my needs as user. (Ponder that...)

**Comment by hartkopp** [comment]
>> Or you may have your desired functionality in your private repo.

> Yip. This is the only viable solution to my needs as user. (Ponder that...)

I did and still do.

candump receives data from the CAN bus.
When you shut down the interface you now get an information about it.
When the interface comes up again you get new CAN frames.

So where's the difference between getting new CAN frames AND getting new CAN frames and know that the interface has gone up 543 ms before?

**Comment by ajneu** [comment]
> So where's the difference between getting new CAN frames AND getting new CAN frames and know that the interface has gone up 543 ms before?

The difference is that in the second case I know that the interface was down and has gone up.

Why do I want this information? Because then when I do long-term tests, with candump, I can look at the log and know that libsocketcan's `can_do_stop` and `can_do_start` (which I happen to use) were called and when they were called. Otherwise I'm just guessing (or must fiddle with my user application, and turn it into a tool, that I expect candump to be)...

**Comment by hartkopp** [comment]
> Because then when I do long-term tests, with candump, I can look at the log and know that libsocketcan's `can_do_stop` and `can_do_start` (which I happen to use) were called and when they were called. Otherwise I'm just guessing (or must fiddle with my user application, and turn it into a tool, that I expect candump to be)...

Yes. Fiddle it into **your** user application and print a proper timestamp when you perform the `can_do_stop` and `can_do_start` actions with it.
https://memegenerator.net/instance/71557527

**Pull request closed without merging**